src/lib/utils/sqlite_wrapper.cpp

```diff
@@ -19,27 +19,222 @@
-      sqlite3_open_v2(":memory:", &_db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX, nullptr);  // NOLINT
+      sqlite3_open_v2(uri.c_str(), &db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX,  // NOLINT
+                      nullptr);
   if (ret != SQLITE_OK) {
-    sqlite3_close(_db);
-    Fail("Cannot open database: " + std::string(sqlite3_errmsg(_db)));
+    sqlite3_close(db);
+    Fail("Cannot open database: " + std::string(sqlite3_errmsg(db)));
   }
 
   // Make LIKE case sensitive, just like in Hyrise
-  raw_execute_query("PRAGMA case_sensitive_like = true");
+  raw_execute_query("PRAGMA case_sensitive_like = true;");
+}
+
+SQLiteWrapper::Connection::Connection(Connection&& other) noexcept : db(other.db) { other.db = nullptr; }
+
+SQLiteWrapper::Connection::~Connection() {
+  if (db) {
+    sqlite3_close(db);
+  }
+}
+
+SQLiteWrapper::Connection SQLiteWrapper::new_connection() const { return Connection{_uri}; }
+
+std::shared_ptr<Table> SQLiteWrapper::Connection::execute_query(const std::string& sql) const {
+  sqlite3_stmt* sqlite_statement;
+
+  auto sql_pipeline = SQLPipelineBuilder{sql}.create_pipeline();
+  const auto& queries = sql_pipeline.get_sql_per_statement();
+
+  // We need to split the queries such that we only create columns/add rows from the final SELECT query
+  std::vector<std::string> queries_before_select(queries.begin(), queries.end() - 1);
+  std::string select_query = queries.back();
+
+  int rc;
+  for (const auto& query : queries_before_select) {
+    rc = sqlite3_prepare_v2(db, query.c_str(), -1, &sqlite_statement, nullptr);
+
+    if (rc != SQLITE_OK) {
+      sqlite3_finalize(sqlite_statement);
+      Fail("Failed to execute query \"" + query + "\": " + std::string(sqlite3_errmsg(db)) + "\n");
+    }
+
+    while ((rc = sqlite3_step(sqlite_statement)) != SQLITE_DONE) {
+    }
+  }
+
+  rc = sqlite3_prepare_v2(db, select_query.c_str(), -1, &sqlite_statement, nullptr);
+
+  if (rc != SQLITE_OK) {
+    auto error_message = "Failed to execute query \"" + select_query + "\": " + std::string(sqlite3_errmsg(db));
+    sqlite3_finalize(sqlite_statement);
+    Fail(error_message);
+  }
+
+  auto result_table = create_hyrise_table_from_result(sqlite_statement, sqlite3_column_count(sqlite_statement));
+
+  if (!sqlite3_stmt_readonly(sqlite_statement)) {
+    // We need to make sure that we do not call sqlite3_reset below on a modifying statement - otherwise, we would
+    // re-execute the modification
+    Assert(!result_table, "Modifying statement was expected to return empty table");
+    sqlite3_finalize(sqlite_statement);
+    return nullptr;
+  }
+
+  sqlite3_reset(sqlite_statement);
+
+  while ((rc = sqlite3_step(sqlite_statement)) == SQLITE_ROW) {
+    copy_row_from_sqlite_to_hyrise(result_table, sqlite_statement, sqlite3_column_count(sqlite_statement));
+  }
+
+  Assert(rc == SQLITE_DONE, "Unexpected sqlite state");
+
+  sqlite3_finalize(sqlite_statement);
```

Comment:
mmh? As sqlite is not an RAII-style C++ library but a C library, we need to finalize things ourselves.

Reply:
Ok, that is good to know. Also found this in the SQLite doc:

> The application must finalize every prepared statement in order to avoid resource leaks. It is a grievous error for the application to try to use a prepared statement after it has been finalized.